### ISSUE TITLE
Support grouped chain WHERE

### ISSUE THREAD

**Issue by techniq** [js-data/js-data-sql#67]
Currently there is no way to group WHERE conditionals with the [query syntax](http://www.js-data.io/docs/query-syntax).  This becomes a problem when using `OR`, for example:

```js
Customer.findAll({
  where: {
    'customer.name': {'|ilike': '%foo'},
    'customer.code': {'|ilike': '%foo'},
    'customer.reseller_id': {'==': 25},
  }
});
```
which produces:
```sql
select "customer".* from "customer"
where
  "customer"."name" ilike '%foo%'
  or "customer"."customer_code" ilike '%foo%'
  and "customer"."reseller_id" = 25
```

but I need to be able to generate the following query:
```sql
select "customer".* from "customer"
where
  ("customer"."name" ilike '%foo%' or "customer"."customer_code" ilike '%foo%')
  and "customer"."reseller_id" = 25
```

One thought I had was to support passing an array of objects to `where`, and each object in the array would be wrapped in parenthesis (using knex's [grouped chain](http://knexjs.org/#Builder-where))
```js
Customer.findAll({
  where: [{
    'customer.name': {'|ilike': '%foo'},
    'customer.code': {'|ilike': '%foo'}
  }, {
    'customer.reseller_id': {'==': 25},
  }]
});
```

@jmdobry thoughts?

**Comment by jmdobry** [comment]
How are paren groups joined? Here you joined the two groups with an `and`, what if we wanted to join the two groups with an `or`?

**Comment by techniq** [comment]
@jmdobry that was one thing I was trying to figure out a good way to handle.

Maybe something like:
```js
Customer.findAll({
  where: {
    'and': [{ // or `or`
      'customer.name': {'|ilike': '%foo'},
      'customer.code': {'|ilike': '%foo'}
    }, {
      'customer.reseller_id': {'==': 25},
    }]
});
```
where an object is items is still `and` together, but a user could be more verbose with their defination:
```js
Customer.findAll({
  where: {
    'and': {
      'and': {
        'customer.name': {'|ilike': '%foo'},
        'customer.code': {'|ilike': '%foo'}
      }
    }, {
      'customer.reseller_id': {'==': 25},
    }
});
```
or even remove the `|` prefix and use:
```js
Customer.findAll({
  where: {
    'and': {
      'or': {
        'customer.name': {'ilike': '%foo'},
        'customer.code': {'ilike': '%foo'}
      }
    }, {
      'customer.reseller_id': {'==': 25},
    }
});
```
---
another idea is to still use the array, but allow operators as strings to be added as well:
```js
Customer.findAll({
  where: [{
    'customer.name': {'|ilike': '%foo'},
    'customer.code': {'|ilike': '%foo'}
  }, 'and', {
    'customer.reseller_id': {'==': 25},
  }]
});
```
I like the terse and readability of this option, but mixing types in an array is a little offputting

**Comment by techniq** [comment]
Implemented in [1.0.0-beta.2](https://github.com/js-data/js-data-sql/commit/842f6b320ae72d04396da1cbad3802cca2c76ff4) - requires js-data v3

**Comment by jmdobry** [comment]
Also implemented in js-data 3.0.0-beta.6
